From 18de11ca2890b7b78faaa83ebde69daafc707230 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Fri, 2 Jun 2023 17:34:16 +0800
Subject: [PATCH 403/780] [Huawei] blk-io-hierarchy: add debugfs for throttle

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188826

--------------------------------

Introduce new debugfs entry to show user io stats about throttled bio:

/sys/kernel/debug/block/[device]/blk_io_hierarchy/throtl/stats

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-io-hierarchy/Makefile   |  1 +
 block/blk-io-hierarchy/debugfs.h  | 32 +++++++++++++++++++++++++++++
 block/blk-io-hierarchy/stats.c    |  4 ++++
 block/blk-io-hierarchy/throttle.c | 34 +++++++++++++++++++++++++++++++
 4 files changed, 71 insertions(+)
 create mode 100644 block/blk-io-hierarchy/debugfs.h
 create mode 100644 block/blk-io-hierarchy/throttle.c

diff --git a/block/blk-io-hierarchy/Makefile b/block/blk-io-hierarchy/Makefile
index d528fd7cf196..e053270972c3 100644
--- a/block/blk-io-hierarchy/Makefile
+++ b/block/blk-io-hierarchy/Makefile
@@ -5,3 +5,4 @@
 obj-$(CONFIG_BLK_IO_HIERARCHY_STATS) += blk_io_hierarchy_stats.o
 
 blk_io_hierarchy_stats-y := stats.o
+obj-$(CONFIG_BLK_DEV_THROTTLING) += throttle.o
diff --git a/block/blk-io-hierarchy/debugfs.h b/block/blk-io-hierarchy/debugfs.h
new file mode 100644
index 000000000000..aef355374b94
--- /dev/null
+++ b/block/blk-io-hierarchy/debugfs.h
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2023. Huawei Technologies Co., Ltd. All rights reserved.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef BLK_IO_HIERARCHY_DEBUGFS_H
+#define BLK_IO_HIERARCHY_DEBUGFS_H
+
+#include "stats.h"
+#include "../blk-mq-debugfs.h"
+
+#define STATS_SHOW_FN(stage, name)				\
+static int							\
+hierarchy_##name##_stats_show(void *data, struct seq_file *m)	\
+{								\
+	blk_io_hierarchy_stats_show(data, m, stage);		\
+	return 0;						\
+}
+
+#ifdef CONFIG_BLK_DEV_THROTTLING
+void io_hierarchy_register_throttle(struct request_queue *q);
+#endif
+
+#endif
diff --git a/block/blk-io-hierarchy/stats.c b/block/blk-io-hierarchy/stats.c
index f63fd675eec1..53dcce8745ec 100644
--- a/block/blk-io-hierarchy/stats.c
+++ b/block/blk-io-hierarchy/stats.c
@@ -14,11 +14,15 @@
 #include <linux/debugfs.h>
 
 #include "stats.h"
+#include "debugfs.h"
 #include "../blk-mq-debugfs.h"
 
 typedef void (register_fn)(struct request_queue *);
 
 static register_fn *register_ops[NR_STAGE_GROUPS] = {
+#ifdef CONFIG_BLK_DEV_THROTTLING
+	[STAGE_THROTTLE] = io_hierarchy_register_throttle,
+#endif
 };
 
 #define io_hierarchy_add(statsp, stage, field, group, nr) \
diff --git a/block/blk-io-hierarchy/throttle.c b/block/blk-io-hierarchy/throttle.c
new file mode 100644
index 000000000000..ac52f0b982aa
--- /dev/null
+++ b/block/blk-io-hierarchy/throttle.c
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2023. Huawei Technologies Co., Ltd. All rights reserved.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/debugfs.h>
+
+#include "debugfs.h"
+
+STATS_SHOW_FN(STAGE_THROTTLE, throttle);
+
+static struct blk_mq_debugfs_attr throtl_hierarchy_debugfs_attrs[] = {
+	{"stats", 0400, hierarchy_throttle_stats_show},
+	{},
+};
+
+void io_hierarchy_register_throttle(struct request_queue *q)
+{
+	struct dentry *dir = debugfs_create_dir(
+			"throtl", q->io_hierarchy_stats->debugfs_dir);
+	if (IS_ERR(dir))
+		return;
+
+	debugfs_create_files(dir, q->io_hierarchy_stats,
+			     throtl_hierarchy_debugfs_attrs);
+}
-- 
2.17.1

